packages/@aws-cdk/aws-glue/lib/job-executable.ts

```diff
@@ -0,0 +1,415 @@
+import * as constructs from 'constructs';
+
+/**
+ * AWS Glue version determines the versions of Apache Spark and Python that are available to the job.
+ *
+ * @see https://docs.aws.amazon.com/glue/latest/dg/add-job.html.
+ *
+ * If you need to use a GlueVersion that doesn't exist as a static member, you
+ * can instantiate a `GlueVersion` object, e.g: `GlueVersion.of('1.5')`.
+ */
+export class GlueVersion {
+  /**
+   * Glue version using Spark 2.2.1 and Python 2.7
+   */
+  public static readonly V0_9 = new GlueVersion('0.9');
+
+  /**
+   * Glue version using Spark 2.4.3, Python 2.7 and Python 3.6
+   */
+  public static readonly V1_0 = new GlueVersion('1.0');
+
+  /**
+   * Glue version using Spark 2.4.3 and Python 3.7
+   */
+  public static readonly V2_0 = new GlueVersion('2.0');
+
+  /**
+   * Glue version using Spark 3.1.1 and Python 3.7
+   */
+  public static readonly V3_0 = new GlueVersion('3.0');
+
+  /**
+   * Custom Glue version
+   * @param version custom version
+   */
+  public static of(version: string): GlueVersion {
+    return new GlueVersion(version);
+  }
+
+  /**
+   * The name of this GlueVersion, as expected by Job resource.
+   */
+  public readonly name: string;
+
+  private constructor(name: string) {
+    this.name = name;
+  }
+}
+
+/**
+ * Runtime language of the Glue job
+ */
+export enum JobLanguage {
+  /**
+   * Scala
+   */
+  SCALA = 'scala',
+
+  /**
+   * Python
+   */
+  PYTHON = 'python',
+}
+
+/**
+ * Python version
+ */
+export enum PythonVersion {
+  /**
+   * Python 2 (the exact version depends on GlueVersion and JobCommand used)
+   */
+  TWO = '2',
+
+  /**
+   * Python 3 (the exact version depends on GlueVersion and JobCommand used)
+   */
+  THREE = '3',
+}
+
+/**
+ * The job type.
+ *
+ * If you need to use a JobType that doesn't exist as a static member, you
+ * can instantiate a `JobType` object, e.g: `JobType.of('other name')`.
+ */
+export class JobType {
+  /**
+   * Command for running a Glue ETL job.
+   */
+  public static readonly ETL = new JobType('glueetl');
+
+  /**
+   * Command for running a Glue streaming job.
+   */
+  public static readonly STREAMING = new JobType('gluestreaming');
+
+  /**
+   * Command for running a Glue python shell job.
+   */
+  public static readonly PYTHON_SHELL = new JobType('pythonshell');
+
+  /**
+   * Custom type name
+   * @param name type name
+   */
+  public static of(name: string): JobType {
+    return new JobType(name);
+  }
+
+  /**
+   * The name of this JobType, as expected by Job resource.
+   */
+  public readonly name: string;
+
+  private constructor(name: string) {
+    this.name = name;
+  }
+}
+
+interface SharedJobExecutableProps {
+  /**
+   * Glue version.
+   *
+   * @see https://docs.aws.amazon.com/glue/latest/dg/release-notes.html
+   */
+  readonly glueVersion: GlueVersion;
+
+  /**
+   * Specifies the Amazon Simple Storage Service (Amazon S3) path to a script that executes a job.
+   */
+  readonly scriptLocation: string;
```

Comment:
the type of this property should be more similar to [lambda.Code](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-lambda.Code.html) – it should allow users to either specify an existing S3 object or bundle and upload one as part of their CDK app

Reply:
That's a good point that I wanted to discuss.
1. As you can see in the different props, there's a `scriptLocation`, `extraJars`, `extraPythonFiles`, etc. i.e. multiple code assets. Should each single item of them be its own object e.g. `glue.Code`? so it becomes
```
scriptLocation: glue.Code,
extraJars: glue.Code[],
```
2. Like `lambda.Code`, the main thing is that the result of this should be able to return an s3 path when bound. As such, I might just begin with the simplest version of just `glue.Code.fromBucket(bucket, key)` and leave the asset handling for a later version maybe?

Reply:
I've taken a stab at https://github.com/aws/aws-cdk/pull/12506/commits/e860f4d31c7abd624c5e88663ddaaa798f9d9595, what do you think? I ended up adding an `AssetCode` as well similar to Lambda to simplify integ tests!